### PR TITLE
🍒[5.7][Distributed] Require that SerReq can only be used with protocols

### PR DESCRIPTION
**Pull request by ktoso**
**Description:** In order to in the future support  non "ad-hoc" protocol requirements we must make it possible to express `T: SerializationRequirement` in the future.

Since the `:` implies being a class or protocol, and currently we only have use cases for protocols, let us ban other things so we can figure it out in the near future.

**Risk:** Low.
**Review by:** @xedin @hborla 
**Testing:** PR validation
**Original PR:** https://github.com/apple/swift/pull/42348
**Radar:** rdar://91663499

### CODE REVIEW

**Comment by ktoso** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckDistributed.cpp [hunk 1]
```diff
@@ -431,6 +431,47 @@ static bool checkDistributedTargetResultType(
...
+  auto &C = nominal->getASTContext();
+  auto DAS = C.getDistributedActorSystemDecl();
+
+  // === AssociatedTypes
+  // --- SerializationRequirement MUST be a protocol TODO(distributed): rdar://91663941
+  // we may lift this in the future and allow classes but this requires more
+  // work to enable associatedtypes to be constrained to class or protocol,
+  // which then will unlock using them as generic constraints in protocols.
+  Type requirementTy = getDistributedSerializationRequirementType(nominal, DAS);
+  requirementTy->dump();
```
Comment (ktoso) [inline lib/Sema/TypeCheckDistributed.cpp:451]: Removing this, was noticed in https://github.com/apple/swift/pull/42420/

**Comment by ktoso** [comment]
@swift-ci please test

File: lib/AST/DistributedDecl.cpp [hunk 2]
```diff
@@ -271,12 +271,24 @@ swift::getDistributedSerializationRequirements(
   if (existentialRequirementTy->isAny())
     return true; // we're done here, any means there are no requirements
 
-  auto serialReqType = existentialRequirementTy->castTo<ExistentialType>()
-                           ->getConstraintType()
-                           ->getDesugaredType();
+  if (auto alias = dyn_cast<TypeAliasType>(existentialRequirementTy.getPointer())) {
+    auto ty = alias->getDesugaredType();
+    if (isa<ClassType>(ty) || isa<StructType>(ty) || isa<EnumType>(ty)) {
```
Comment (hborla) [inline lib/AST/DistributedDecl.cpp:276]: There is still some unaddressed review feedback from the `main` PR related to this part of the change: https://github.com/apple/swift/pull/42348#discussion_r852211348

Can you please address the feedback?

Comment (ktoso) [reply lib/AST/DistributedDecl.cpp:276]: Sorry on the huge delay here and thanks for pointing out I missed that!

Back to compiler now post-WWDC and will fix here and on main 👍

Comment (ktoso) [reply lib/AST/DistributedDecl.cpp:276]: Thanks @hborla  -- I followed up here https://github.com/apple/swift/pull/59404 could you have a look and if it's correct I'll apply here as well 🙏

**Review by Gam3356 — COMMENTED** [review]
They are hacking files need help

File: lib/AST/DistributedDecl.cpp (hunk 2 above)
Comment (ktoso) [reply lib/AST/DistributedDecl.cpp]: Added the cleanup on main and here as well, thank you for noticing @hborla !

**Comment by ktoso** [comment]
@swift-ci please test

**Review by hborla — APPROVED** [review]

**Pull request merged**